File: 03 network/tcpip/echo_srv.py
#!/usr/bin/env python
#-*- coding: utf-8 -*-
#If this code works, it was written by Seunghyun Lee(www.bluebaynetworks.co.kr).
#If not, I don't know who wrote it

import thread
import socket

TCP_PORT = 5005
BUFFER_SIZE = 1024  

#에코 클라이언트와 통신하는 쓰레드
def echo_job(conn):
	while 1:
		data = conn.recv(BUFFER_SIZE)
		if not data: break
		print "received data:", data
		conn.send(data)  # echo
	conn.close()
	print "client exit"

#통신에 사용할 TCP/IP 소켓 생성
s = socket.socket(socket.AF_INET, socket.SOCK_STREAM)

try:
	#통신에 사용할 포트 번호를 5005로 지정. 접속 IP에는 제한을 두지 않음
	s.bind(('', TCP_PORT))
	print "TCP Socket bind success"
	#에코 클라이언트의 접속을 기다림
	s.listen(1)
	print "TCP Socket waits for client"
except socket.error, exc:
	#bind 에러 발생시 에러코드 및 내용을 보여주고 프로그램을 종료
	print "Caught exception socket.error : %s" % exc
	exit(0)

try: 
	while 1:
		conn, addr = s.accept()
		print 'New Echo Client connection address:', addr
		#에코 클라이언트와 통신할 새로운 쓰레드를 만들고 다시 새로운 접속을 기다림
		thread.start_new_thread(echo_job, (conn,))
except (KeyboardInterrupt, SystemExit):
	print("Exit...")
finally:
	s.close()

File: 03 network/tcpip/echo_srv.c
// echo_srv.cpp : Defines the entry point for the console application.
//


#include <stdio.h>
#include <errno.h>
#include <fcntl.h>
#include <sys/types.h>
#include <sys/socket.h>
#include <resolv.h>
#include <signal.h>
#include <pthread.h>
#include<string.h>


int g_nPort = 5005;


void *echo_job(void *psock)
{
	int bytes, x;
	char buffer[1024];
	int sock = *((int *)psock);

	while(1)
	{
		memset(buffer, 0x00, sizeof(buffer));
		bytes = read(sock, buffer, sizeof(buffer));
		if ( bytes <= 0 ) break;
		printf("received data:%s\n", buffer);
		write(sock, buffer, bytes + 1);
	}
	printf("client exit\n");
	close(sock);
	return NULL;
}



int main(int argc, char* argv[])
{
	int sd, thr_id, client;
	struct sockaddr_in addr, cli;
	pthread_t p_thread;
	socklen_t size;

	size = sizeof(cli);
	sd = socket(PF_INET, SOCK_STREAM, 0);
	bzero(&addr, sizeof(addr));
	addr.sin_family = AF_INET;
	addr.sin_port = htons(g_nPort);
	addr.sin_addr.s_addr = INADDR_ANY;
	if ( bind(sd, (struct sockaddr*)&addr, sizeof(addr)) != 0 )
		perror("bind()");

	listen(sd, 15);
	for (;;)
	{
		client = accept(sd, (struct sockaddr*)&cli, &size);
		if ( client > 0 )
		{
			printf("Connection address:%s\n", inet_ntoa(cli.sin_addr));
			thr_id = pthread_create(&p_thread, NULL, echo_job, (void *)&client);
		}
	}
	close(sd);
	return 0;
}

